### ISSUE TITLE
OpenEMR Cloud - v1 Beta AWS Guide

### ISSUE THREAD

**Issue by MatthewVita** [openemr/openemr#606]
## WHO

[Team Epsilon](https://github.com/GoTeamEpsilon/purpose). We are a small open source development group working on interesting projects that serve as educational resources as well as provide value to existing open source projects.

## WHAT

Our team will provide a turnkey solution for small facilities and hospitals to run their OpenEMR v5 installation in the AWS cloud. This solution will be peformant, scalable, secure (HIPAA-compliant), and cost-effective.

Many OpenEMR users host their server instances on premise and have not yet realized the benefits of cloud technologies. While users can hire a professional vendor for cloud deployment services or simply learn AWS and do it themselves, our team has identified a need for a more cost effective option.

## WHY

Our team just shipped our [last project](https://github.com/GoTeamEpsilon/angular-to-react-redux) and are looking to start working on the next cool thing!

## HOW

The first release (codename "Dove") will use AWS services in a way that can be described as a "PaaS-buffet". This is to say that the chosen technologies will "lock" the user to many opinionated AWS services. Services include Elastic Beanstalk, RDS, Route53, IAM, TurnKey CouchDB, Cloudwatch, CloudFormation, and SES.

The output of the project will conclude with documentation+scripts being checked into OpenEMR so that users can do a "one click install"!

## ETC

- This issue is just a placeholder for the work that is happening on the Team Epsilon repo.
- If you wish to contribute to this project, message/email me
- This falls under the category of "Modernization" work
- 2 cloud experts are on this project... should be moving quickly once we get started up!

**Comment by MatthewVita** [comment]
UPDATE: I received a message in regards to if this work has for-profit motives. This is not the case. Team Epsilon is a group of friends building open source solutions in their free time.

This is a solution for OpenEMR users to have a "one click install" OpenEMR cloud instance without hassle.

**Comment by MatthewVita** [comment]
UPDATES:

- We have an AWS account for our group. Gotta watch those $$$s!
- We are working out _which_ AWS services are going to be used.
- We are working on testing out CouchDB in the cloud! Very interesting task.

**Comment by MatthewVita** [comment]
- Version 1 will not used CloudFormation (will just be a manual series of steps as a soft launch). Automation will be done in CloudFormation in Version 1.5 to truly allow a user to "turn the key".
- Our team has figured out the architecture:

![image](https://cloud.githubusercontent.com/assets/1715206/25065840/8e0da048-21e4-11e7-945c-be5cb265b693.png)

**Comment by bobinson** [comment]
for PHP sessions we have elastic cache - similarly, are there any shared files ? (I am not very familiar with the system as such.)

ie, smarty templates, any cache files, images etc. ie unless all the files and images wrt a specific location is not mapped to a single instance, there may be missing files etc ?

**Comment by bradymiller** [comment]
Note sure if this is related, but there is an issue here discussing migration of writable files/dirs to a centralized location:
https://github.com/openemr/openemr/issues/572

**Comment by MatthewVita** [comment]
@bobinson Shared files are possible through CouchDB. By default, OpenEMR puts documents on disk. However, CouchDB can be easily configured to be the centralized cluster. This is going to be rough on AWS because there isn't any turnkey solution available. We're going to have to set it up as Docker images in ECS, most likely.

@bradymiller, @bobinson:

- openemr/sites/default/documents (will be stored in CouchDB, no issue)
- openemr/sites/default/edi (may be specific to any given instance, is an issue)
- openemr/sites/default/era (may be specific to any given instance, is an issue)
- openemr/sites/default/letter_templates (may be specific to any given instance, is an issue)
- openemr/gacl/admin/templates_c (may be specific to any given instance, is an issue)
- openemr/interface/main/calendar/modules/PostCalendar/pntemplates/compiled (may be specific to any given instance, is an issue)
- openemr/interface/main/calendar/modules/PostCalendar/pntemplates/cache (may be specific to any given instance, is an issue)

-m

**Comment by bobinson** [comment]
@MatthewVita and @bradymiller   - I am not familiar with the structure of the software. Will play around a bit and revert here. 

The above list is helpful.

> This is going to be rough on AWS because there isn't any turnkey solution available 


In the past (some 9 years back, before S3) I had used NFS (yea, SPOF and I am not very proud about it) :)

**Comment by MatthewVita** [comment]
@bobinson Just had an "ah ha!" moment. NFS! That must be the solution... https://aws.amazon.com/efs/ is what I was looking for!

**Comment by MatthewVita** [comment]
UPDATES:

- Our team is happy with that architecture graphic above... we're running forward with it!
- We are going to cheat and do a version `1.0.0-ALPHA` that isn't HIPAA compliant for users outside of the US. The official v1 will be HIPAA compliant, but we want to get something out of the door. Reason for this is EFS isn't HIPAA compliant out of the box and we'll have to patch a good bit of stuff manually to get that to work. Also, Daniel and I aren't 100% sure of the AWS Config Rules that have to be applied for HIPAA.

**Comment by MatthewVita** [comment]
UPDATES:

- I put together a contrived PHP/Apache application to simulate OpenEMR and the `sites/default` documents using EFS... working on two EC2 instances in ELB: https://github.com/GoTeamEpsilon/OpenEMR-AWS-Guide/blob/master/Other/MRV_working_guide.MD w00t!
- Daniel is studying HIPAA compliance and various AWS security bits

**Comment by MatthewVita** [comment]
UPDATES:

- Daniel is diving deep into AWS VPCs
- I am working on a proof of concept for `$_SESSION`s to be stored in AWS Elasicache (more difficult than you might think!)

**Comment by MatthewVita** [comment]
UPDATES:

- Daniel is setting up IAM roles
- Jeffrey is looking into AWS AMI Linux security
- I am fighting with putting `$_SESSION`s in elasticache. I've reached out to some AWS experts with the following question:

*Objective*

Store PHP7 `$_SESSION['...']` information (from an ElasticBeanstalk app) in a centralized Memcached cluster via AWS Elasticache.

*Manual Setup Steps*

0. (here will be steps on how to create an ElasticBeanstalk application and the user will be asked to note the security group that it is using)
1. In the AWS console, navigate to Services > ElastiCache and click "Get Started Now"
2. Under "Cluster engine" check "Memcached"
3. Under "Memcached settings" enter "sessions" for "Name" and leave all default settings (will be more specific after I'm done with this proof of concept)
4. Now on the Dashboard, wait until "status" is "created" and then expand the row and note the public dns (note that steps will be added to add the security group from step 0 will be here as well... I am doing this now btw)
5. On your local machine, create a directory called `sample`
6. Under `sample`, create `.ebextensions/elasticcache-sessions.config` with the following content:

```
files:
  "/etc/php.d/project.ini" :
    mode: "000644"
    owner: root
    group: root
    content: |
      [Session]
      session.save_handler = memcached
      session.save_path = "tcp://dns-noted-from-step-4:11211"
```

7. Still under `sample`, create `index.php` with the following content:

```
<?php
    header('Content-Type: text/plain');
    session_start();
    if(!isset($_SESSION['visit']))
    {
        echo "This is the first time you're visiting this server\n";
        $_SESSION['visit'] = 0;
    }
    else
            echo "Your number of visits: ".$_SESSION['visit'] . "\n";

    $_SESSION['visit']++;

    echo "Server IP: ".$_SERVER['SERVER_ADDR'] . "\n";
    echo "Client IP: ".$_SERVER['REMOTE_ADDR'] . "\n";
    print_r($_COOKIE);
```

8. Compress the `sample` directory as a `.zip` file and deploy it to the Beanstalk application noted in step 0
9. Visit <dns-for-beanstalk>/sample/index.php in your browser and... 


*Problem*

`PHP Fatal error:  session_start(): Failed to create session ID: memcached.` is thrown. If I SSH into an EC2 instance belonging to Beanstalk, I can see that PHP7 is installed. If I `php --ini | grpe memcached`, there are results. I also am able to verify that the ebextension I wrote is getting applied. Ready to bang my head against the wall here :slightly_smiling_face:. Your help is appreciated.

**Comment by MatthewVita** [comment]
UPDATES:

- After _many_ hours of hard work, I pivoted and solve the $_SESSION problem by using Redis. See my final solution here: https://github.com/GoTeamEpsilon/OpenEMR-AWS-Guide/blob/master/Other/MRV_working_guide.MD#redis-sessions

- Because I am lazy, please see the copy/paste below to get an idea of what everyone is working on:

![image](https://cloud.githubusercontent.com/assets/1715206/26026829/dafdb936-37d0-11e7-9b43-366cd2fa2f38.png)

**Comment by MatthewVita** [comment]
UPDATES:

A super productive weekend! The actual guide portion is coming together. See https://github.com/GoTeamEpsilon/OpenEMR-AWS-Guide for a sneak peak.

**Comment by MatthewVita** [comment]
UPDATES:

- Daniel is looking into Route53
- I just finished off the last 10% of the ebextensions... https://github.com/GoTeamEpsilon/OpenEMR-AWS-Guide/tree/master/assets/eb (currently requires user to copy/paste values in, but I have a ticket in Asana to make it all environment variable based)

**Comment by MatthewVita** [comment]
UPDATES:

- I'm working on the general administration guide (...what happens when stuff hits the fan and I need to check logs? How do I deploy some code changes?...)
- Daniel is doing great work with Route53 research
- Small note, but it is neat to see our project is 100% shell coding:

![image](https://cloud.githubusercontent.com/assets/1715206/26533377/25be3132-43e6-11e7-9b91-35d3e2854964.png)

Elite hackers only!

**Comment by MatthewVita** [comment]
UPDATES:

- Project is moving on strong. Here are the remaining tasks for the beta:

![1](https://cloud.githubusercontent.com/assets/1715206/26662029/53461ca8-464f-11e7-98ae-3e0893200647.png)

- And the "stable" release will be the turnkey solution with cloud formation:

![2](https://cloud.githubusercontent.com/assets/1715206/26662043/64758acc-464f-11e7-9926-eb4e5c8abfdf.png)

**Comment by MatthewVita** [comment]
UPDATES:

- Things are still going well despite some bugs with Redis sessions.
- We added a new team member!
- v1 beta will be shipped around July 15th!

**Comment by MatthewVita** [comment]
UPDATES:

- Recruited new team members Shivam, Brian, Stephen, and Saurav
- Redis sessions are now working perfectly!
- Matthew is working on getting EFS to mount correctly and in a reliable way (kind of a challenge!)
- OpenEMR install and update scripts are now hidden to the public internet after initial install

**Comment by MatthewVita** [comment]
UPDATES:

- I got EFS to mount correctly and in a reliable way (big bug fix!)
- Saurav is working on setting up the Docker containers for compose and eventually Kubernetes

**Comment by MatthewVita** [comment]
UPDATES:

- The README now points to our 3 ongoing solutions. The starred one is the manual approach which will be released as a beta solution for outside of the US:

![image](https://user-images.githubusercontent.com/1715206/27254356-c426fb9c-5354-11e7-9078-96ace6761bcf.png)

- The beta guide is now in a handy "book" format:

![image](https://user-images.githubusercontent.com/1715206/27254360-d542dfae-5354-11e7-8a08-f392876762b3.png)

![image](https://user-images.githubusercontent.com/1715206/27254367-e6e4c43e-5354-11e7-84d7-5b58d1d0e70b.png)

- The first pass through of an architecture diagram is checked in:

![architecture](https://user-images.githubusercontent.com/1715206/27254377-009eb876-5355-11e7-8a94-6328f1a3f337.png)

**Comment by bradymiller** [comment]
wow!

**Comment by MatthewVita** [comment]
UPDATES:

The beta manual guide is almost ready for testing...

![image](https://user-images.githubusercontent.com/1715206/27412678-354fd24a-56c4-11e7-9915-55e2b9f70785.png)

While we have 2 team members working on the stable turnkey version of the project, the beta will be great for users to test. I'm very excited to move away from the manual guide because it's error prone and complex. Automation is the key and that is the team's goal after the beta launch.

**Comment by MatthewVita** [comment]
UPDATE:

- The team is deciding if v1 stable should use CloudFormation or Terraform. Terraform is an open source technology that can deploy systems to any cloud. For the case of AWS, it uses the non-free CloudFormation stuff under the hood if I understand correctly. This will be useful for v2, which will use Docker/Kubernetes.
- Almost done with v1 beta guide:

![image](https://user-images.githubusercontent.com/1715206/27510675-b6f51bea-58e3-11e7-836d-48f89c3f1531.png)

**Comment by MatthewVita** [comment]
UPDATES:

- Version 1 Beta is coming along. A friend of mine is interested in a career in QA and he will be QAing the guide.
- Version 1 Stable has been defined... we're going with Terraform!

![untitled](https://user-images.githubusercontent.com/1715206/27566870-f891d36c-5ab5-11e7-8ce0-df2c0a295d2d.png)

**Comment by MatthewVita** [comment]
UPDATES:

- Version 1 Beta is very close:

![image](https://user-images.githubusercontent.com/1715206/27713252-c200861e-5cf8-11e7-8298-00f058beef56.png)

@bradymiller I don't think it's ready to be checked into OpenEMR (I was thinking it could get it's own folder under `Documentation`) until these 2 tasks are complete, but it may be a great time to ask users if they'd like to try it out! Do you think it would make sense to simply link to https://github.com/GoTeamEpsilon/OpenEMR-Cloud/blob/master/v1-Beta-AWS-Guide/README.md and ask for folks to run through the guide or do you think I should put together a production environment that can be linked to for folks to play around on?

Thanks,
Matthew

**Comment by bradymiller** [comment]
That manual is pretty awesome. Also functions as a very nice practical tutorial on using the cloud.

Definitely feel free to link to this from forum/wiki (and twitter/facebook if wish) to see if users wish to try it out. I wouldn't feel too inclined to need to set up a production environment; I'll leave that decision up to you :)

**Comment by MatthewVita** [comment]
Hi Brady,

Glad you found the manual useful. I'm _so_ looking forward to v1 stable where it's all automated.

I will actually hold off on posting this to the community until these 2 tasks are done. They are important and will make a big difference in the experience!

Thanks,
Matthew

**Comment by MatthewVita** [comment]
ETA is sometime next week because Daniel will be free to finish his task then.

**Comment by andremillet** [comment]
lol I need production working. still don't have a clue - and now, even time - on how to install it for tomorrow

**Comment by MatthewVita** [comment]
Hi @andremillet, the good news is that my team will have a "soft" release of the v1 Beta AWS Guide this week! There will be a Pull Request made to get the guide checked into the official OpenEMR repository.

I will most definitely communicate the updates to you as progress is made.

Thanks,
Matthew

**Comment by andremillet** [comment]
​you rock, buddy !​


*André Millet*


2017-07-04 12:31 GMT-03:00 Matthew Vita <notifications@github.com>:

**Comment by MatthewVita** [comment]
UPDATES:

- As seen below, CloudTrail chapter is done, but CloudWatch is being finished up. Brian is doing a great job here.
- The video and the upstream commit tickets are really just placeholders. Not too much work involved there.
- Asher kindly offered his expertise on security and locked everything behind a VPN. He put together a rather large PR that I am reviewing now. While this wasn't originally in scope, it would be crazy to skip out on it because it's such a security enhancement! I a reviewing/testing it now.

![image](https://user-images.githubusercontent.com/1715206/27888145-89c052ac-61b2-11e7-84a6-fa235efa24ee.png)

**Comment by andremillet** [comment]
please, allow me to understand nothing, but trust you guys fully 😁

**Comment by MatthewVita** [comment]
@andremillet Yeah, it's better that way to be honest. I'll join you. This stuff is insanely complex. Lot to keep in mind!

**Comment by MatthewVita** [comment]
First pass at the VPN solution was reviewed/tested/tweaked and checked into the guide. There needs to be many light changes to wording to make it more clear and consistent and the admin chapter is somewhat out of date, but this was a HUGE win for the project.

Thanks to Asher!

**Comment by MatthewVita** [comment]
BIG UPDATE:

https://github.com/GoTeamEpsilon/OpenEMR-Cloud/releases/tag/v1-Beta-AWS-Guide is done! Going to change around the links and remove anything unneeded and put together a PR for OpenEMR's "Documentation" folder. Will do that tomorrow.

**Comment by bradymiller** [comment]
Nice!

**Comment by andremillet** [comment]
😁👏🏽👏🏽👏🏽👏🏽👏🏽

**Comment by MatthewVita** [comment]
UPDATES:

- Asher, Daniel, and Brian are really doing great work on the automated solution. Although it isn't multicloud at the moment (it is AWS-specific), this may be a great v1 stable release. We can move the originally planned v1 stable release with multi-cloud support to v2 and the originally planned v2 release with Kubernetes to v3. For my own sanity, I will be closing this ticket out shortly and creating new tickets here to map and track each project. TL;DR: v1 stable (automated and HIPAA compliant solution) may be done within the next couple of weeks if we keep up our momentum! I am so fortunate to have some real DevOps experts working on the project!

- Anyone following this thread should "star" https://github.com/OpenVPN/openvpn. It is what we are using for tunneling into the cloud and I am extremely impressed. This is the best VPN solution I've ever used and most of the solutions I've used in the past were non-free and had people working on them full time...

PLAN FOR V1 BETA:

- I am going to run through the guide to create @andremillet's cloud environment today. This will not only be a great test of the guide from end to end, but will mean that we officially have a beta user!

- I will be forking the v1 beta to trim it down to just what needs to be checked in for the OpenEMR `documentation` folder. I have licensed the guide as MIT, so there should be 0 conflicts with checking it in because MIT is compatible inside of a GPL project. The PR to @bradymiller, but I don't expect that you test it as that would take a long time :).
